### DISCUSSION TITLE
Initializing HeliumFluidProperties within another Class

### DISCUSSION THREAD

**Discussion by MikeLeimon** [idaholab/moose#22899] [Q&A Modules: General] [answered]
I've been working on trying to leverage the `HeliumFluidProperties` class that is included within MOOSE within a custom class that I am creating. Initially, I tried including a copy of the HFP object as a reference in the protected section my new class: (i.e., `HeliumFluidProperties & _hfp`). While this was what I would prefer, I couldn't figure out how to correctly define the value within the class contructor. I was trying to do something like the following: `_hfp(declareProperty<HeliumFluidProperties>("hfp", parameters))`, however this never worked...

So after looking at that for a while, I gave up on only constructing the object once and decided to just create an HFP object in a function right before I needed to use it. Based on the unit test included in MOOSE, I settled upon doing something like the following:

```c++
  HeliumFluidProperties hfp = HeliumFluidProperties(
    HeliumFluidProperties::validParams());
```

While this code would finally compile, it did cause the MOOSE application to crash and burn during execution. Included below is a small section of the errors I ended up with as the code was crashing:

```
ERROR: no MooseApp* parameter named "_moose_app" found.

Known parameters:
Name	 Type	 Value
---------------------
 T_initial_guess	 double	 400
 _aux_sys	 AuxiliarySystem*	 0
 _eigen_problem	 EigenProblem*	 0
 _executioner	 Transient*	 0
 _fe_problem	 FEProblem*	 0
 _fe_problem_base	 FEProblemBase*	 0
 _material_data_type	 Moose::MaterialDataType	 0
 _moose_base	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 FluidProperties
 _moose_warehouse_system_name	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 UserObject
 _nl_sys	 SystemBase*	 0
 _object_name	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 
 _residual_object	 bool	 0
 _subproblem	 SubProblem*	 0
 _sys	 SystemBase*	 0
 _tid	 unsigned int	 0
 _type	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 
 _unique_name	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 
 allow_duplicate_execution_on_initial	 bool	 0
 allow_imperfect_jacobians	 bool	 0
 control_tags	 std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >	 
 enable	 bool	 1
 execute_on	 ExecFlagEnum	 TIMESTEP_END
 force_postaux	 bool	 0
 force_preaux	 bool	 0
 force_preic	 bool	 0
 fp_type	 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >	 single-phase-fp
 p_initial_guess	 double	 200000
 prop_getter_suffix	 MaterialPropertyName	 
 tolerance	 double	 1e-08
 use_displaced_mesh	 bool	 0

Stack frames: 35
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: /usr/share/moose/framework/libmoose-opt.so.0(+0x10f75b2) [0x7fdd4284e5b2]
3: MooseApp* InputParameters::getCheckedPointerParam<MooseApp*>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const
4: MooseObject::MooseObject(InputParameters const&)
5: UserObject::UserObject(InputParameters const&)
6: GeneralUserObject::GeneralUserObject(InputParameters const&)
7: ThreadedGeneralUserObject::ThreadedGeneralUserObject(InputParameters const&)
8: FluidProperties::FluidProperties(InputParameters const&)
9: SinglePhaseFluidProperties::SinglePhaseFluidProperties(InputParameters const&)
10: HeliumFluidProperties::HeliumFluidProperties(InputParameters const&)
```

So what I would really like to know is how I should go about initializing this fluid property within another custom class. Is it best to include it along with the other members of the class and initialize it at the same time, or would the second approach I took be more appropriate, or is the correct answer something else entirely.

Anyhow, thanks in advance for any advice/help you guys provide me with!

**Comment by cpgr** [comment, marked as answer]
I think that `BrineFluidProperties` does something similar to what you want, in that it constructs `WaterFluidProperties` and `NaClFluidProperties` internally to have access to those methods. Maybe you could find inspiration there?

**Reply by GiudGiud** [reply]
Agree with that suggestion

  HeliumFluidProperties hfp = HeliumFluidProperties(
    HeliumFluidProperties::validParams());
does not work because while these are valid parameters, they are empty.
It s better to pass the current parameters, the ones coming in the constructor of your object.
They may need additional parameters set if your object has less parameters than the ones you are trying to create within your object

**Reply by MikeLeimon** [reply]
Thank you very much! 

Looking to `BrineFluidProperties` for inspiration did allow me to come up with some code that worked.
